### ISSUE TITLE
[Project]: Add Automatic Music Genre Classification using Mel-Spectrograms and CNN (GTZAN Dataset)

### ISSUE THREAD

**Issue by divyanshim27** [abhisheks008/DL-Simplified#1200]
## Summary

I would like to propose and implement: **Automatic Music Genre Classification using Mel-Spectrograms and a CNN.** This is a classic and well-loved audio deep learning problem that is accessible to beginners while covering important concepts (spectrogram generation, 2D CNN on audio, transfer learning). DL-Simplified currently has almost no audio classification projects beyond a single Cats vs Dogs audio project. Adding a music genre classifier introduces contributors to:
- Short-Time Fourier Transform (STFT) and mel-spectrogram generation
- Treating audio as a 2D image classification problem
- Transfer learning with `EfficientNet-B0` applied to audio spectrograms

## Proposed Implementation

**Dataset:** GTZAN Music Genre Dataset — 1,000 audio tracks, 30 seconds each, across 10 genres (blues, classical, country, disco, hiphop, jazz, metal, pop, reggae, rock). Available on Kaggle under CC BY-NC 4.0.

**Feature extraction:**
```python
import librosa
import librosa.display
import numpy as np

def extract_mel_spectrogram(file_path: str, duration: float = 30.0,
                             n_mels: int = 128, hop_length: int = 512) -> np.ndarray:
    """
    Load audio and convert to mel-spectrogram.
    Returns array of shape (n_mels, time_frames) normalized to [0, 1].
    """
    y, sr = librosa.load(file_path, duration=duration, mono=True)
    mel = librosa.feature.melspectrogram(y=y, sr=sr, n_mels=n_mels,
                                          hop_length=hop_length, fmax=8000)
    mel_db = librosa.power_to_db(mel, ref=np.max)
    # Normalize to [0, 1]
    return (mel_db - mel_db.min()) / (mel_db.max() - mel_db.min())
```

**Three models compared in the notebook:**

1. **Custom CNN from scratch:**
```python
class GenreCNN(nn.Module):
    def __init__(self, num_classes=10):
        super().__init__()
        self.features = nn.Sequential(
            nn.Conv2d(1, 32, 3, padding=1), nn.BatchNorm2d(32), nn.ReLU(),
            nn.MaxPool2d(2),
            nn.Conv2d(32, 64, 3, padding=1), nn.BatchNorm2d(64), nn.ReLU(),
            nn.MaxPool2d(2),
            nn.Conv2d(64, 128, 3, padding=1), nn.BatchNorm2d(128), nn.ReLU(),
            nn.AdaptiveAvgPool2d((4, 4))
        )
        self.classifier = nn.Linear(128 * 4 * 4, num_classes)
```

2. **EfficientNet-B0 fine-tuned** (spectrograms treated as 3-channel images via channel replication)

3. **MFCC + Random Forest** (classical ML baseline for comparison)

**Expected Results:** EfficientNet-B0 achieves ~87–91% test accuracy on GTZAN.

**Notebook structure:**

Music Genre Classification using DL/
├── Dataset/
│ └── README.md # GTZAN Kaggle link + preprocessing
├── Images/
│ ├── mel_spectrogram_blues.png
│ ├── mel_spectrogram_metal.png
│ ├── model_comparison_accuracy.png
│ └── confusion_matrix_efficientnet.png
├── Model/
│ ├── Music_Genre_Classification.ipynb
│ └── README.md
├── Web App/
│ ├── app.py # Gradio app: upload .wav → predicted genre
│ └── README.md
└── requirements.txt

Could you please assign this issue to me?

**Labels:** `enhancement`, `new-project`, `audio-deep-learning`, `intermediate`, `help wanted`, `GSSoC 2026`
